src/catalog/src/expr_cache.rs

```diff
@@ -0,0 +1,594 @@
+                let mut exp = (GlobalId::User(next_id), generate_expressions());
+                next_id += 1;
+                exp.1.optimizer_features = optimizer_features.clone();
+                cache
+                    .insert_expressions(vec![exp.clone()], BTreeSet::new())
+                    .await;
+                current_ids.insert(exp.0);
+                exps.insert(exp.0, exp.1);
+            }
+            exps
+        };
+
+        {
+            // Re-open the cache at the first generation.
+            let (_cache, entries) =
+                ExpressionCacheHandle::spawn_expression_cache(ExpressionCacheConfig {
+                    deploy_generation: first_deploy_generation,
+                    persist,
+                    organization_id,
+                    current_ids,
+                    optimizer_features,
+                    remove_prior_gens,
+                })
+                .await;
+            assert_eq!(
+                entries, exps,
+                "Previous generation expressions should still exist"
+            );
+        }
+
+        {
+            // Open the cache at a new generation and clear previous generations.
+            remove_prior_gens = true;
+            let (_cache, entries) =
+                ExpressionCacheHandle::spawn_expression_cache(ExpressionCacheConfig {
+                    deploy_generation: second_deploy_generation,
+                    persist,
+                    organization_id,
+                    current_ids,
+                    optimizer_features,
+                    remove_prior_gens,
+                })
+                .await;
+            assert_eq!(
+                entries, new_gen_exps,
+                "new generation expressions should be persisted"
+            );
+        }
+
+        {
+            // Re-open the cache at the first generation.
+            let (_cache, entries) =
+                ExpressionCacheHandle::spawn_expression_cache(ExpressionCacheConfig {
+                    deploy_generation: first_deploy_generation,
+                    persist,
+                    organization_id,
+                    current_ids,
+                    optimizer_features,
+                    remove_prior_gens,
+                })
+                .await;
+            assert_eq!(
+                entries,
+                BTreeMap::new(),
+                "Previous generation expressions should be cleared"
+            );
+        }
+    }
+
+    proptest! {
+        #![proptest_config(ProptestConfig::with_cases(32))]
+
+        #[mz_ore::test]
+        #[cfg_attr(miri, ignore)]
+        fn expr_cache_roundtrip((key, val) in any::<(CacheKey, Expressions)>()) {
+            let (encoded_key, encoded_val) = ExpressionCodec::encode(&key, &val);
+            let (decoded_key, decoded_val) = ExpressionCodec::decode(encoded_key, encoded_val);
+
+            assert_eq!(key, decoded_key);
+            assert_eq!(val, decoded_val);
+        }
+    }
```

Comment:
FYI, this test with just 32 cases is pretty slow. It's possible that `Arbitrary` creates completely unrealistically huge expressions, but I am worried about the serialization perf in prod.